### ISSUE TITLE
[Help] Stop Release enforced unnecessarily?

### ISSUE THREAD

**Issue by nephros** [Olf0/sfos-upgrade#97]
**SailfishOS VERSION** : 2.0.0.10
<br />**HARDWARE** : Jolla Tablet
<br />**sfos-upgrade VERSION**: 3.9.17
<br />

#### QUESTION

Internal list of stop releases: Is sfos-upgrade more zealous than the Jolla list?

#### STEPS TO REPRODUCE

Running sfos-upgrade on 2.0.0.10, wanting to update to 2.2.0.29 will output:

    Notice: Upgrading from 2.0.0.10 to 2.2.0.29 would omit installing 2.0.5.6 as a stop release!
 
Now, the d.s.o page on [Releases](https://docs.sailfishos.org/Support/Releases/) sais this step, 2.0.0.10 -> 2.2.0.29 should be possible.

I don't mind sfos-update being more careful than the list prescribes, but I was curious whether there was maybe some outdated internal list in the tool, or an omission on Jolla's list.

#### ADDITIONAL INFORMATION

By the way, sfos-upgrade 3.9.17 *is* capable of performing its duty on SFOS 1.1.9.30 which was a bit surprising to me but kudos to you for making it so very portable.

And more by the way, actually sfos-upgrade does give an error when running it on such old releases, namely:

```
[root@Jolla nemo]# sfos-upgrade 2.2.0.29
/usr/bin/sfos-upgrade: line 205: sailfishdocs_sfos_releases: unbound variable
Warning: "2.2.0.29" does not seem to be a publicly released SailfishOS version!
Do you really want to continue? (y/N) y
```

I haven't investigated deeply but I assume it's simply because ancient ssl certificates on such old releases will make the `curl` call(s) fail. Harmless in my book so I'm ignoring that.

**Comment by Olf0** [comment]
* Thank you for reporting.  Had CoViD the first time, hence the delay.
* Yes, the aim was and is, the one can update a factory reset Jolla 1 with it, starting at SFOS 1.0.0.1 or even 0.99.x.y.  This was even tested a few times ... long ago.
* WRT 2.0.5.6 handled as a stop release, see https://github.com/Olf0/sfos-upgrade/blob/master/usr/bin/sfos-upgrade#L223-L224 which provides https://together.jolla.com/question/207172/downgrade-sailfish-os/?answer=207190#post-id-207190
* WRT `[root@Jolla nemo]# sfos-upgrade 2.2.0.29`
  `/usr/bin/sfos-upgrade: line 205: sailfishdocs_sfos_releases: unbound variable`
  `Warning: "2.2.0.29" does not seem to be a publicly released SailfishOS version!`
  This is an error, hence I reopen this issue.

**Comment by Olf0** [comment]
> Hope you have recovered well.
> 
> On the topic of "weird" stop releases, I believe 4.5.0.16 is not a stop release any more, but the script treats it as such.

@nephros, you are regressing :wink:: as always the point release number is irrelevant WRT stop releases, traversing any point release of a stop release will do.

**Comment by Olf0** [comment]
Fixed by commit [0427d48](https://github.com/Olf0/sfos-upgrade/commit/0427d48da15677d6a69c461c404f7741985f40a6).

**Comment by Olf0** [comment]
> [On the topic of "weird" stop releases, I believe 4.5.0.16 is not a stop release any more, but the script treats it as such.](https://github.com/Olf0/sfos-upgrade/issues/97#issuecomment-1454653160)

BTW, as so often Jolla misses to update their lists, e.g. [docs.sailfishos.org/Support/Releases/README.md](https://github.com/sailfishos/docs.sailfishos.org/blob/master/Support/Releases/README.md) still [displays 4.5.0.16 as latest stop release as of 2023-03-18](https://github.com/sailfishos/docs.sailfishos.org/commit/24c796b094a22d15d2abbdc20c4442a4adc536e8), despite [4.5.0.18 was GA-released on 2023-02-16](https://forum.sailfishos.org/t/release-notes-struven-ketju-4-5-0-18/14656#version-history-of-the-os-release-1).  Hence Jolla regularly tricks `sfos-upgrade` into using an older point release as a stop release, despite me having updated `sfos-upgrade`'s internal list of stop releases in a timely manner (usually within days after a new point release of a stop release).

Consequently I had no choice, but to rework and complicate the way the stop releases are handled: Now they are stripped of their point release number and then matched against the highest point release of each SailfishOS version X.Y.Z.  See PR #103 for details.
